functions/src/pdfBuilder/donationCertificateBuilderFunction.ts

```diff
@@ -0,0 +1,140 @@
+import * as functions from 'firebase-functions';
+//import { createAnduploadDonationCertificate } from './buildDonationCertificateCH';
+import { initializeApp } from "firebase/app";
+import { getFirestore, connectFirestoreEmulator } from 'firebase/firestore';
+import { doc, setDoc, getDocs, collection } from "firebase/firestore";
+import { createAndUploadDonationCertificate } from './generateDonationCertificatePDF';
+import * as config from '../config';
+
+let FIRESTORE_EMULATOR_REQUIRED = true;
+
+export const bulkDonationCertificateBuilderFunction = functions.https.onCall(async (data, context) => {
+	const app = await initializeFirebaseApp();
+	const firestore = await initializeFirestore(app);
+	const now = new Date();
+	const currentDate = now.toLocaleDateString('de-DE');
+
+	const calculationYear = data.year
+
+	for (let i = 0; i < data.users.length; ++i) {
```

Comment:
shall we use a for loop?
```
for await (const user of data.users ) {
}
```

Reply:
Already changed. @andrashee, I would be interested why this is better than for loop

Reply:
In this context, I think it doesn't really matter and is a more cosmetic question which one to prefer. I find the syntax of `for each` the more modern and appealing way, but feel free to use whatever you like more.

Generally, I guess one tries to abstract away implementation details (e.g. in this case how to access elements in a data structure) as much as possible. This allows easier modularization and reuse of methods.

Your implementation assumes that `data.users` is an indexed data structure (e.g. array), but for the use case it actually only needs to be iterable. E.g. if you have a function like:
```
const createAndUploadDonationCertificate = async (user: User) => {{}}

const createAndUploadDonationCertificates = async (users: Iterable<User>) => {
	for await (const user of users) {
		await createAndUploadDonationCertificate(user);
	}
}
```

Then you can feed it with whatever data structure implements the iterable interface, and you delegate the implementation details of looping to the data structure. E.g. index based for arrays as your implementation, pointers to the next entries for linked lists and streams, ... 
For example, if we all of a sudden would have many users, one could then easily switch to read a user stream from firestore instead of loading the entire array into memory to keep the memory pressure O(1) instead of O(users). 

But this more as a high-level explanation, not that it matters in our case here :-).